Add async method code generation

antoyo opened the pull request:
For now, it's usable (tested with gio, gtk and sourceview), but we could improve the heuristic as mentionned in the TODO comment.

The other two things that will be implemented in next PRs will be the annotation to specify what is the finish function when it cannot be guessed and writing Unimplemented next to Cancellable when this type does not exist (i.e. when the user forgets to add it to `Gir.toml`).

I believe some code should be improved, especially the trampoline code generation.
If you have any other improvements, I'd be glad to hear them.

antoyo commented:
cc @GuillaumeGomez @EPashkin @sdroege

GuillaumeGomez commented on src/analysis/bounds.rs:
```diff
@@ -99,21 +102,42 @@ impl Bounds {
+    ) -> (Option<String>, Option<String>) {
+        let type_name = bounds_rust_type(env, par.typ);
+        let mut type_string =
+            if async {
```
I suppose you could write it like this:

```
if async && async_param_to_remove(&par.name) {
    return (None, None);
} else {
    type_name.into_string()
}
```

GuillaumeGomez commented on src/analysis/bounds.rs:
```diff
@@ -149,7 +173,23 @@ impl Bounds {
+                        info_for_next_type: false,
+                    });
+                    return true;
+                } else {
```
This `else` condition is unnecessary since the `if` returns.

GuillaumeGomez commented on src/analysis/bounds.rs:
```diff
@@ -196,18 +236,20 @@ impl Bounds {
-            .find(move |n| if n.parameter_name == name {
+            .find(move |n| {
+                if n.parameter_name == name {
                 !n.info_for_next_type
```
The indent is broken in here. :p

GuillaumeGomez commented on src/analysis/function_parameters.rs:
```diff
@@ -161,9 +166,15 @@ pub fn analyze(
+            library::ParameterDirection::Out => !can_as_return(env, par) && !async_func,
         };
 
+        if async_func {
```
```
if async_func && async_param_to_remove(&par.name) {
    add_rust_parameter = false;
}
```

GuillaumeGomez commented on src/analysis/rust_type.rs:
```diff
@@ -241,11 +241,13 @@ pub fn parameter_rust_type(
     let type_ = env.library.type_(type_id);
     let rust_type = rust_type_full(env, type_id, nullable, ref_mode);
     match *type_ {
+        Fundamental(super::library::Fundamental::Pointer) if async => Ok("TODO4".to_string()), // FIXME: remove
```
Wut?

antoyo replied:
Yeah, some TODO that are not generated.
I'll have to figure out why.

GuillaumeGomez commented:
What I'm seeing for the moment seems correct (even if it seems incomplete). However I'm not the maintainer in here hehe.

sdroege commented:
Can you create a PR for gio with the changes this would add, i.e. all the implementations for the async methods in GFile?

EPashkin commented on src/codegen/function_body_chunk.rs:
```diff
@@ -128,12 +120,100 @@ impl Builder {
+                .map(|param| (param, type_mem_mode(env, param)))
+                .map(|(param, mode)| format!("let mut {} = {};\n", param.name, mode_to_string(mode)))
+                .collect::<Vec<_>>()
+                .join("                ");
```
Please change it to "\t\t\t\t"

antoyo replied:
I'll instead change that to use `Chunk` directly, to avoid having code as strings.

EPashkin commented on src/codegen/function_body_chunk.rs:
```diff
@@ -128,12 +120,100 @@ impl Builder {
+                        Ok(({result}))
+                    }} else {{
+                        Err(from_glib_full(error))
+                    }};
```
IMHO `let result = ` better be oneliner.

EPashkin commented on src/codegen/function_body_chunk.rs:
```diff
@@ -431,3 +511,75 @@ impl Builder {
+    if id == namespaces::MAIN {
+        return "ffi".to_string();
+    }
+    format!("{}_ffi", name)
```
`Namespace.ffi_crate_name` is not same?

antoyo replied:
`ffi_crate_name` is only available on the other `Namespace` struct.

EPashkin commented:
@antoyo Overall is good idea, I slightly not like "internal trampoline" but it not so big problem. Thanks.

antoyo commented:
@sdroege I'll do it when I get back to Canada, so maybe in one week.
@EPashkin I didn't know how to do it (not sure if I can reuse the trampoline logic), so I did it this way because I wanted to have something working during the HackFest.
But the overall trampoline code generation should be redone before we merge this PR.

Thanks to all of you for your comments.

EPashkin commented:
@antoyo don't worry about trampolines.

antoyo commented:
@sdroege Here's the [PR](https://github.com/gtk-rs/gio/pull/62).

antoyo commented:
I've fixed all the issues you mentioned and I've fixed the async trampoline code generation.
I think this PR is ready to be tested.
Do we have any infrastructure to test `gir` against every GNOME library that we support?

EPashkin commented:
@antoyo No we don't have :(
I tested building of examples on updated gtk, it builds fine.

EPashkin commented:
@GuillaumeGomez we add async function it this release?

sdroege commented:
I added some comments about the generated code here https://github.com/gtk-rs/gio/pull/63

sdroege commented:
IMHO this is good to go, judging from the generated code.

GuillaumeGomez commented:
Seems good to me as well. Is it good for you @EPashkin or do you want to add something else @antoyo?

EPashkin commented:
@GuillaumeGomez Its good for me too.
You wan't add this before release?

GuillaumeGomez commented:
Yes if possible so I'll handle the release tomorrow as the latest (won't be able to do it before next week-end otherwise).

antoyo commented:
@sdroege @GuillaumeGomez I wanted to fix the `Cancellable` issue mentionned by @sdroege.
I think I can do it today.

antoyo commented:
This will require to add `Gio.Cancellable` in `Gir.toml` files when needed.
I'll commit the fix soon.

EPashkin commented:
Gtk-rs examples still builds.
@sdroege, @antoyo there still any problem? Or I can merge?

GuillaumeGomez commented:
For me it was already good so as you wish!

antoyo commented:
It's ready to be merged.
Just waiting for @sdroege to see the latest change (`&&Fn`).

sdroege commented:
See my comment there

antoyo commented:
I've fixed this issue in the latest commit.

antoyo commented:
@sdroege I've fixed the other issues mentionned in the gio PR.

EPashkin commented:
@antoyo Thanks.
I add regen PR for sourceview. Gtk, glib has only commented changes and IMHO can be ignored.

EPashkin commented:
Actually gtk also have async functions in `IconInfo`, but current version not works good outside `gio` as it don't add `use gio_ffi`

EPashkin commented:
Created fix, will add regen for Gtk too.

sdroege commented:
Looks good to me, just need to add support for the "progress-callback" pattern in the future, as used by various async operations
